### PR TITLE
Node repurpose tool

### PR DESCRIPTION
**Pull request by henningandersen**
When a node is repurposed to master/no-data or no-master/no-data, v7.x
will not start (see #37748 and #37347). The `elasticsearch-node repurpose`
tool can fix this by cleaning up the problematic data.

Unsure if this should be backported to 7.0 too?

Documentation to be done in a follow-up PR.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+import java.util.stream.IntStream;
+
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.not;
+import static org.mockito.Matchers.contains;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class NodeRepurposeCommandIT extends ESIntegTestCase {
```
Comment (andrershov) [inline server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java]: Since it's an integration test, I think we need to be able to verify that the document that we have indexed is lost. For example, you may start one master-only node and one data-only node and then repurpose data-only to be master-only or no-master no-data. In this case, when the repurposed node joins the cluster, the index will be re-created but the document is gone.  
Also due to randomization, `executeRepurposeCommand` is a bit hard to follow. I think here we can stop a particular node using `nameFilter` and explicitly run repurpose tool successfully on this node and unsuccessfully on another node.

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    }
+
+    private void verifyUnchangedDataFiles(CheckedRunnable<? extends Exception> runnable) throws Exception {
+        long before = digestPaths();
+        runnable.run();
+        long after = digestPaths();
+        assertEquals("Must not touch files", before, after);
+    }
+
+    private long digestPaths() {
```
Comment (andrershov) [inline server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java]: I wonder if we already have a similar function implemented somewhere (it takes 40 lines of code here) or probably it makes sense to add it to some util class. Probably @ywelsch can suggest something.

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    public void testRepurpose() throws Exception {
```
Comment (andrershov) [inline server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java]: Since we have tests for `RepurposeCommand` not in `ElasticsearchNodeCommandIT` can we rename `ElasticsearchNodeCommandIT` to make it clear that it tests unsafe bootstrap and detach cluster?

File: server/src/main/java/org/elasticsearch/cluster/coordination/NodeToolCli.java [hunk 4]
```diff
@@ -36,6 +37,7 @@ public NodeToolCli() {
         CommandLoggingConfigurator.configureLoggingWithoutConfig();
         subcommands.put("unsafe-bootstrap", new UnsafeBootstrapMasterCommand());
         subcommands.put("detach-cluster", new DetachClusterCommand());
+        subcommands.put("repurpose", new NodeRepurposeCommand());
     }
 
     public static void main(String[] args) throws Exception {
```
Comment (andrershov) [inline server/src/main/java/org/elasticsearch/cluster/coordination/NodeToolCli.java]: The description that we pass to super when constructing NodeToolCli no longer looks correct.
Don't forget about `ArchiveTestCase` that tests this message

**Review by andrershov — CHANGES_REQUESTED** [review]
@henningandersen thanks for your PR! I've left several comments, overall looks good already.

File: server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    protected void processNodePaths(Terminal terminal, Path[] dataPaths, Environment env) throws IOException {
+        assert DiscoveryNode.isDataNode(env.settings()) == false;
+
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java]: Fixed in 18df8d71aa3d1ec27090005cad512e238c3c3692

File: server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+        Set<Path> indexPaths = uniqueParentPaths(shardDataPaths, indexMetaDataPaths);
+        if (indexPaths.isEmpty()) {
+            terminal.println(Terminal.Verbosity.NORMAL, NO_DATA_TO_CLEAN_UP_FOUND);
+            return;
+        }
+
+        Set<String> indexUUIDs = indexUUIDsFor(indexPaths);
+        outputVerboseInformation(terminal, nodePaths, indexPaths, indexUUIDs);
+
+        terminal.println(noMasterMessage(indexUUIDs.size(), shardDataPaths.size(), indexMetaDataPaths.size()));
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java]: I intentionally had the longer list of indices and shards before the summary, so I kept that but moved "Use -v to see list of paths and indices affected" to be output after the summary.

See 18df8d71aa3d1ec27090005cad512e238c3c3692

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java (hunk 3 above)
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java]: Fixed in 4df583c2cf7ed901b0ee6f20e156c7e70f85e855

File: server/src/main/java/org/elasticsearch/cluster/coordination/NodeToolCli.java (hunk 4 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/cluster/coordination/NodeToolCli.java]: Changed it to be more generic to cover all 3 tools:
`A CLI tool to do unsafe cluster and index manipulations on current node`

See 85cf0e65edcaefb2f80280dd936b7c24839e0eae

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java (hunk 1 above)
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java]: Regarding checking for lost data, I added a check that we loose the index completely in f5396b18519b6d8115852673b43fe692dade310c , I think that should suffice, do you agree?

Commit 85cf0e65edcaefb2f80280dd936b7c24839e0eae fixes test to stop one node, run the tool successfully on that node and unsuccessfully on the other node.

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+        String shardDataDirName = Integer.toString(randomInt(10));
+
+        try (NodeEnvironment env = new NodeEnvironment(settings, TestEnvironment.newEnvironment(settings))) {
+            IndexMetaData.FORMAT.write(IndexMetaData.builder(INDEX.getName())
+                .settings(Settings.builder().put("index.version.created", Version.CURRENT))
+                .numberOfShards(1)
+                .numberOfReplicas(1)
+                .build(), env.indexPaths(INDEX));
+            for (Path path : env.indexPaths(INDEX)) {
+                if (includeShardData) {
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java]: Fixed in 18df8d71aa3d1ec27090005cad512e238c3c3692

File: server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+public class NodeRepurposeCommand extends ElasticsearchNodeCommand {
+
+    private static final Logger logger = LogManager.getLogger(NodeRepurposeCommand.class);
+
+    static final String ABORTED_BY_USER_MSG = ElasticsearchNodeCommand.ABORTED_BY_USER_MSG;
+    static final String FAILED_TO_OBTAIN_NODE_LOCK_MSG = ElasticsearchNodeCommand.FAILED_TO_OBTAIN_NODE_LOCK_MSG;
+    static final String NO_CLEANUP = "Node has node.data=true -> no clean up necessary";
+    static final String NO_DATA_TO_CLEAN_UP_FOUND = "No data to clean-up found";
+    static final String NO_SHARD_DATA_TO_CLEAN_UP_FOUND = "No shard data to clean-up found";
+    static final String PRE_V7_MESSAGE =
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java]: Fixed in 18df8d71aa3d1ec27090005cad512e238c3c3692

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+    private Settings settings;
+    private Environment environment;
+    private Path[] nodePaths;
+    private Settings dataNoMasterSettings;
+    private Settings noDataNoMasterSettings;
+    private Settings noDataMasterSettings;
+
+    @Before
+    public void createNodePaths() throws IOException {
+        settings = buildEnvSettings(Settings.EMPTY);
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java]: Fixed in 18df8d71aa3d1ec27090005cad512e238c3c3692

File: server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+            pathsToCleanup.forEach(p -> terminal.println(Terminal.Verbosity.VERBOSE, "  " + p.toString()));
+            terminal.println(Terminal.Verbosity.VERBOSE, "Indices affected:");
+            indexUUIDs.forEach(uuid -> terminal.println(Terminal.Verbosity.VERBOSE, "  " + toIndexName(nodePaths, uuid)));
+        } else {
+            terminal.println("Use -v to see list of paths and indices affected");
+        }
+    }
+
+    private String toIndexName(NodeEnvironment.NodePath[] nodePaths, String uuid) {
+        Index index = new Index("dummy", uuid);
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java]: Fixed in 18df8d71aa3d1ec27090005cad512e238c3c3692

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java (hunk 2 above)
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java]: I found none, will await feedback from @ywelsch on this.

**Comment by henningandersen** [comment]
Thanks for a thorough review, @andrershov , I believe I addressed all comments, so is ready for another round.

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java (hunk 1 above)
Comment (andrershov) [reply server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java]: I'm not sure how the test works, but after you restart two nodes - they should not form the cluster, because there is no longer the master node. It'd nice to figure out why ensureGreen returns successfully in this case.
But in general, I think, having one dedicated master-eligible node and one data node and repurposing of data node (mentioned in the comment) would be easier to follow.

File: server/src/main/java/org/elasticsearch/cluster/coordination/NodeToolCli.java (hunk 4 above)
Comment (andrershov) [reply server/src/main/java/org/elasticsearch/cluster/coordination/NodeToolCli.java]: This description once again makes me think that detach/unsafe bootstrap and repurpose should be two different tools. However, let's leave it as it is for now.

**Review by andrershov — CHANGES_REQUESTED** [review]
Thanks for the changes @henningandersen. I've left a comment about the test and would like to review it again after it's addressed.

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java (hunk 1 above)
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java]: @andrershov this works because `InternalTestCluster` maintains the voting configuration in `InternalTestCluster.excludeMasters`, which takes the stopping node of the voting configuration. This is used by several other tests (is the default for `InternalTestCluster`), for instance I think `GatewayIndexStateIT.testIndexDeletionWhenNodeRejoins` depends on the same.

Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java]: Changed to one master only and one data only node in 7a0712c9e2f6df80e83d7158ce99b27cf50c88c3.

**Comment by henningandersen** [comment]
> Thanks for the changes @henningandersen. I've left a comment about the test and would like to review it again after it's addressed.

Thanks @andrershov , I modified the test, please have another look.

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java (hunk 1 above)
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java]: Sorry, the test needed more adjustments, fixed in 4cbf288dab44ebb7422437fcccefeb67790797a6

I somehow ran the wrong test 100 times ;-)

**Comment by henningandersen** [comment]
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

**Comment by henningandersen** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java [hunk 11]
```diff
@@ -68,29 +68,25 @@ public void testRepurpose() throws Exception {
...
 
         logger.info("--> Restarting and repurposing other node");
 
         internalCluster().stopRandomNode(s -> true);
         internalCluster().stopRandomNode(s -> true);
 
-        executeRepurposeCommandForOrdinal(noMasterNoDataSettings, indexUUID, 0);
+        executeRepurposeCommandForOrdinal(noMasterNoDataSettings, indexUUID, 0, 0);
 
         internalCluster().startMasterOnlyNode();
```
Comment (andrershov) [inline server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java]: This is not easy to follow. We've repurposed both nodes as non-master and non-data. And now we're starting them as "master" and "data" correspondingly. 
I think, that we at least need to add a comment that describes what we're doing here. 
Like "now both nodes are coordinating, let's start one of them as a master node and the second one as the data node, this does not require repurpose tool invocation. However, since they were converted to coordinating nodes previously, all data and metadata should be lost and we can verify it".

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        internalCluster().stopRandomNode(s -> true);
+
+        executeRepurposeCommandForOrdinal(noMasterNoDataSettings, indexUUID, 0, 0);
+
+        internalCluster().startMasterOnlyNode();
+        internalCluster().startDataOnlyNode();
+
+        ensureGreen();
+
+        // indexes gone.
```
Comment (andrershov) [inline server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java]: index is gone

**Review by andrershov — APPROVED** [review]
@henningandersen I've left several comments to the test, but no need for the next pass.
Since there is no feedback from @yannick on "digest" function, let's have this custom implementation for now.
As for backporting to 7.0, on the one hand, it would be nice to have it in 7.0, but I'm affraid we don't have enough time for documentation. So I think backport to 7.1 is enough.
@yannick what do you think?

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java (hunk 2 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java]: I'm not aware of any such existing functionality. There is `BufferedChecksumIndexInput` in Lucene, but that will not shorten the code here.

**Comment by ywelsch** [comment]
Given that #37748 and #37347 are in 7.0, we should try to get this in there as well. I'm sure we will figure out the docs before release.
